pkg/netpol/connlist/conns_formatter_dot.go

```diff
@@ -0,0 +1,125 @@
+func getPeerLine(peer Peer) (string, bool) {
+	peerName, peerColor, isExternalPeer := peerNameAndColorByType(peer)
+	linePrefix := "\t\t"
+	if isExternalPeer {
+		linePrefix = "\t"
+	}
+	return fmt.Sprintf("%s%q [label=%q color=%q fontcolor=%q]", linePrefix, peerName, peerName, peerColor, peerColor), isExternalPeer
+}
+
+// returns a dot string form of connections from list of Peer2PeerConnection objects
+func (d formatDOT) writeOutput(conns []Peer2PeerConnection) (string, error) {
+	nsPeers := make(map[string][]string)         // map from namespace to its peers (grouping peers by namespaces)
+	externalPeersLines := make([]string, 0)      // list of peers which are not in a cluster's namespace (will not be grouped)
+	edgeLines := make([]string, len(conns))      // list of edges lines
+	peersVisited := make(map[string]struct{}, 0) // acts as a set
+	for index := range conns {
+		srcStr, dstStr := conns[index].Src().String(), conns[index].Dst().String()
+		edgeLines[index] = getEdgeLine(conns[index])
+		if _, ok := peersVisited[srcStr]; !ok {
+			peersVisited[srcStr] = struct{}{}
+			externalSrcLine := checkAndAddPeerToNsGroup(nsPeers, conns[index].Src())
+			if externalSrcLine != "" {
+				externalPeersLines = append(externalPeersLines, externalSrcLine)
+			}
+		}
+		if _, ok := peersVisited[dstStr]; !ok {
+			peersVisited[dstStr] = struct{}{}
+			externalDstLine := checkAndAddPeerToNsGroup(nsPeers, conns[index].Dst())
+			if externalDstLine != "" {
+				externalPeersLines = append(externalPeersLines, externalDstLine)
+			}
+		}
+	}
+	// sort graph lines
+	sort.Strings(edgeLines)
+	sort.Strings(externalPeersLines)
+	// collect all lines by order
+	allLines := []string{common.DotHeader}
+	allLines = append(allLines, addNsGroups(nsPeers)...)
+	allLines = append(allLines, externalPeersLines...)
+	allLines = append(allLines, edgeLines...)
+	allLines = append(allLines, common.DotClosing)
+	return strings.Join(allLines, newLineChar), nil
+}
+
+// checks if the peer is in cluster's namespace, then adds its line to the namespace list in the given map.
+// else, returns its line to be added to the external peers lines
+func checkAndAddPeerToNsGroup(mapNsToPeers map[string][]string, peer Peer) string {
+	peerLine, isExternalPeer := getPeerLine(peer)
+	if !isExternalPeer { // belongs to a cluster's namespace
+		if _, ok := mapNsToPeers[peer.Namespace()]; !ok {
+			mapNsToPeers[peer.Namespace()] = []string{}
+		}
+		mapNsToPeers[peer.Namespace()] = append(mapNsToPeers[peer.Namespace()], peerLine)
+		return ""
+	}
+	// else case - an external (ip/ ingress-controller) peer
+	return peerLine
+}
+
+func addNsGroups(nsPeersMap map[string][]string) []string {
+	res := []string{}
+	// sort namespaces (map's keys) to ensure same output always
+	nsKeys := sortMapKeys(nsPeersMap)
+	// write ns groups
+	for _, ns := range nsKeys {
+		peersLines := nsPeersMap[ns]
+		sort.Strings(peersLines)
+		// create ns  subgraph cluster
+		nsLabel := strings.ReplaceAll(ns, "-", "_")
```

Comment:
why do we need this `replace`? does it change an original ns name?

Reply:
no, it does not change the ns name (and the original namespace name is displayed in the frame `nsLines = append(nsLines, "\t\tlabel=\""+ns+"\"", "\t}")` which is on line 111),

this `nsLabel with replace`  is only used for writing the `subgraph header` which exists only in the dot file and will not appear in the graph itself.
replaced since the `dot` format does not accept `-` in its sub-graphs names (headers) . (we'll see a format error in the dot editor)

for example running :
`./bin/k8snetpolicy list --dirpath tests/minimal_test_in_ns/ -o dot` 
we have a namespace `hello-world`
in the dot format we'll see :
```
subgraph cluster_hello_world {      --> subgraph header with the replace to avoid dot format error
                "workload-a[Deployment]" [label="workload-a[Deployment]" color="blue" fontcolor="blue"]
                "workload-b[Deployment]" [label="workload-b[Deployment]" color="blue" fontcolor="blue"]
                label="hello-world"    --> here is the actual namespace name (also displayed in the frame)
}
```